src/Npgsql/TypeHandlers/CompositeHandlers/CompositeMemberInfoOfStruct`2.cs

```diff
@@ -0,0 +1,91 @@
+﻿using System;
+using System.Diagnostics;
+using System.Linq.Expressions;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+using Npgsql.PostgresTypes;
+using Npgsql.TypeHandling;
+
+namespace Npgsql.TypeHandlers.CompositeHandlers
+{
+    sealed class CompositeMemberInfoOfStruct<TComposite, TMember> : CompositeMemberInfo<TComposite>
+    {
+        delegate TMember GetMember(ref TComposite composite);
+        delegate void SetMember(ref TComposite composite, TMember value);
+
+        readonly GetMember _get;
+        readonly SetMember _set;
+        readonly NpgsqlTypeHandler _handler;
+
+        public CompositeMemberInfoOfStruct(FieldInfo fieldInfo, PostgresType postgresType, NpgsqlTypeHandler handler)
+            : base(fieldInfo, postgresType)
+        {
+            var composite = Expression.Parameter(typeof(TComposite), "composite");
+            var value = Expression.Parameter(typeof(TMember), "value");
+
+            _get = Expression
+                .Lambda<GetMember>(Expression.Field(composite, fieldInfo), composite)
+                .Compile();
+            _set = Expression
+                .Lambda<SetMember>(Expression.Assign(Expression.Field(composite, fieldInfo), value), composite, value)
+                .Compile();
+            _handler = handler;
+        }
+
+        public CompositeMemberInfoOfStruct(PropertyInfo propertyInfo, PostgresType postgresType, NpgsqlTypeHandler handler)
+            : base(propertyInfo, postgresType)
+        {
+            var getMethod = propertyInfo.GetGetMethod();
+            if (getMethod?.IsPublic == true)
+                _get = (GetMember)Delegate.CreateDelegate(typeof(GetMember), getMethod);
+
+            var setMethod = propertyInfo.GetSetMethod();
+            if (setMethod?.IsPublic == true)
+                _set = (SetMember)Delegate.CreateDelegate(typeof(SetMember), setMethod);
+
+            _handler = handler;
+        }
+
+        public override async ValueTask Read(object composite, NpgsqlReadBuffer buffer, bool async)
```

Comment:
At least in theory it's possible to accept a `ref TComposite composite`, avoiding the boxing inside ByReference. You can't have `ref` in async methods, but you *could* call all the async methods (e.g. `buffer.Ensure()`) and check whether the Tasks returned completed synchronously. If all the tasks completed synchronously (which they will in most cases) - great, we don't need to box.

But this would get tricky - if we can't complete synchronously we must box (copy the given ref TComposite into an object in the heap), and then we must return that object from the method. The API would look pretty weird but it could work... Also, this would probably break the abstraction that currently exists over struct and struct, although that doesn't seem very important.

What do you think? Am I missing something?

Reply:
Thanks for the idea, will try it.

Reply:
I remember why I didn't go this way. As you said, it won't look nice in that case. Let's do some boxing now? (: If we decide that it hurts performance to much, we can change the code.

By the way, I can optimize this code by making to `Read` methods which will accept `TComposite` and `ByReference<TComposite>` respectively. This change will eliminate one cast in the `Set` method. Don't know is it worth enough.

Reply:
Is it really that ugly/complicated? On the one hand, I think the vast majority of cases would return synchronously, and so benefit from the elimination of the allocation. On the other hand, it's true that composite use in general isn't that widespread, and mapping structs to composites is probably even rarer.

If you really feel like this will make everything too ugly, I'm OK with leaving it out for now. But I might want to take a look myself to see what that involves etc...

Reply:
It's possible only for the sync way, but for other it's impossible becuase nobody know how fields are long. I have an idea, but will try it later.